includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -485,11 +490,50 @@ public function process_payment_for_order( $args ) {
 
 			$intent_id = $intent->get_id();
 			$status    = $intent->get_status();
+			$charge_id = $intent->get_charge_id();
+		} else {
+			// For $0 orders, we need to save the payment method using a setup intent.
+			$intent = $this->payments_api_client->create_setup_intent(
+				$payment_method,
+				$customer_id,
+				'true'
+			);
+
+			$intent_id = $intent['id'];
+			$status    = $intent['status'];
+			$charge_id = '';
 
+			// In SCA cases the setup intent status might be requires_action and we should display the authentication modal.
+			// For now, since we're not supporting SCA cards, we can ignore that status.
+			if ( 'succeeded' !== $status ) {
+				throw new Exception( __( 'Failed to add the provided payment method. Please try again later', 'woocommerce-payments' ) );
+			}
+		}
+
+		if ( ! empty( $intent ) ) {
 			if ( 'succeeded' !== $status && 'requires_capture' !== $status ) {
 				$intent_failed = true;
 			}
 
+			if ( $save_payment_method && ! $intent_failed ) {
+				try {
+					$token = $this->token_service->add_payment_method_to_user( $payment_method, $user );
+				} catch ( Exception $e ) {
+					// If saving the token fails, log the error message but catch the error to avoid crashing the checkout flow.
+					Logger::log( 'Error when saving payment method: ' . $e->getMessage() );
+				}
+			}
+
+			$order_tokens = $order->get_payment_tokens();
+			if ( ! empty( $token ) && $token->get_id() !== end( $order_tokens ) ) {
+				$order->add_payment_token( $token );
```

Comment:
This could lead to the same token being added to the list more than one time, but it is needed so that shoppers may re-use a previous card for the same subscription.

I tried to remove the previous entry for the card, but `WC_Order` does not support it (and [doesn't intend to](https://github.com/woocommerce/woocommerce/issues/11857)), so we'd need to manually interact with the order's `data_store` to do it. I tried modifying it, but it was a bit quirky and did not work perfectly.

Should I dig more into this, or is the current behavior ok, given that reusing the card is more of an edge case?

Reply:
> This could lead to the same token being added to the list more than one time, but it is needed so that shoppers may re-use a previous card for the same subscription.

It is unfortunate, but I also think it's just a nuisance. I would note the unwanted but apparently innocuous behavior in a comment and move on

Reply:
This was added to #826 in https://github.com/Automattic/woocommerce-payments/pull/826/commits/466b85e8cd3e016d4b30ac4469ea4d0e7185e93d and I added a comment in https://github.com/Automattic/woocommerce-payments/pull/832/commits/d3a4890025ad1838b0cffdcc296ad4f36a823c50.